test/e2e/cluster_upgrade.go

```diff
@@ -0,0 +1,134 @@
+		var webserver *WebserverTest
+
+		BeforeEach(func() {
+			framework.beforeEach()
+			webserver = NewWebserverTest(framework.Client, framework.Namespace.Name, name)
+			webserver.CreateWebserverPod()
+			svc := webserver.BuildServiceSpec()
+			svc.Spec.Type = api.ServiceTypeLoadBalancer
+			webserver.CreateService(svc)
+		})
+
+		AfterEach(func() {
+			framework.afterEach()
+			webserver.Cleanup()
+		})
+
+		It("by kube-push of master should maintain responsive services", func() {
+			testClusterUpgrade(framework, name, func() {
+				runUpgradeScript("hack/e2e-internal/e2e-push.sh", "-m")
+			})
+		})
+
+		It("by gce-upgrade of master should maintain responsive services", func() {
+			if !providerIs("gce") {
+				By(fmt.Sprintf("Skippingt test, which is not implemented for %s", testContext.Provider))
+				return
+			}
+			testClusterUpgrade(framework, name, func() {
+				runUpgradeScript("hack/e2e-internal/e2e-upgrade.sh", "-M", "-l")
+			})
+		})
+	})
+})
+
+func testClusterUpgrade(framework Framework, svcName string, upgrade func()) {
+	result, err := waitForLoadBalancerIngress(framework.Client, svcName, framework.Namespace.Name)
+	Expect(err).NotTo(HaveOccurred())
+	if len(result.Status.LoadBalancer.Ingress) != 1 {
+		Failf("got unexpected number (%v) of ingress points for externally load balanced service, only expecting one: %v", result.Status.LoadBalancer.Ingress, result)
+	}
+	ingress := result.Status.LoadBalancer.Ingress[0]
+	ip := ingress.IP
+	if ip == "" {
+		ip = ingress.Hostname
+	}
+
+	By("Waiting for pod to become reachable")
+	testLoadBalancerReachable(ingress, 80)
+
+	httpClient := http.Client{Timeout: 2 * time.Second}
+	done := make(chan struct{}, 1)
+	// Let's make sure we've finished the heartbeat before shutting things down.
+	var wg sync.WaitGroup
+	go util.Until(func() {
+		defer GinkgoRecover()
+		wg.Add(1)
+		defer wg.Done()
+
+		expectNoError(wait.Poll(poll, singleCallTimeout, func() (bool, error) {
+			r, err := httpClient.Get("http://" + ip)
+			if err != nil {
+				return false, nil
+			}
+			if r.StatusCode < http.StatusOK || r.StatusCode >= http.StatusNotFound {
+				return false, nil
+			}
+			return true, nil
+		}))
+	}, 200*time.Millisecond, done)
+
+	By("starting upgrade")
+	upgrade()
+	done <- struct{}{}
+	wg.Wait()
+	Logf("upgrade complete")
+
+	By("validating post upgrade state")
+	pods, err := framework.Client.Pods(framework.Namespace.Name).List(labels.Everything(), fields.Everything())
+	Expect(err).NotTo(HaveOccurred())
+	Expect(len(pods.Items) == 1).Should(BeTrue())
+	_, err = framework.Client.Services(framework.Namespace.Name).Get(svcName)
+	Expect(err).NotTo(HaveOccurred())
+}
+
+func runUpgradeScript(scriptPath string, args ...string) {
+	cmd := exec.Command(path.Join(testContext.RepoRoot, scriptPath), args...)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	if err := cmd.Run(); err != nil {
+		Failf("Upgrade failed: %v", err)
+	}
+	time.Sleep(5 * time.Second)
```

Comment:
I was thinking that I wanted to make sure things didn't come crashing down after the new master came up (like if it came up without and pods triggering the kubelet to kill all the pods that were running so obediently while the master was down). Not sure if that test is neccessary but i'll figure out how to do it without a sleep.